graspologic/nominate/VNviaSGM.py

```diff
@@ -0,0 +1,205 @@
+        Nx2 = np.array(ego_list(BB, self.ell, list(Sx2), mindist=0))
+
+        foo = np.concatenate((Sx1, voi_reord))
+        ind1 = np.concatenate((Sx1, voi_reord, np.setdiff1d(Nx1, foo)))
+        ind2 = np.concatenate((Sx2, np.setdiff1d(Nx2, Sx2)))
+
+        AA_fin = AA[np.ix_(ind1, ind1)]
+        BB_fin = BB[np.ix_(ind2, ind2)]
+
+        seeds_fin = list(range(len(Sx1)))
+        sgm = GMP(n_init=self.R, shuffle_input=False, init_method="rand", padding="naive")
+        corr = sgm.fit_predict(AA_fin, BB_fin, seeds_A=seeds_fin, seeds_B=seeds_fin)
+        P_outp = sgm.probability_matrix_
+
+        self.P = P_outp
+        self.corr = corr
+        self.a_inds = a_reord[ind1]
+        self.b_inds = b_reord[ind2]
+        self.n_seeds_used = len(Sx1)
+
+        nomination_list = list(zip(self.b_inds[self.n_seeds_used:], self.P[0]))
+        nomination_list.sort(key=lambda x: x[1], reverse=True)
+        self.nomination_list = nomination_list
+
+    def fit_predict(self, A, B, seeds_A, seeds_B):
+        """
+        Fits the model with two assigned adjacency matrices, returning optimal
+        permutation indices
+
+        Parameters
+        ----------
+        A : 2d-array, square
+            A square adjacency matrix
+
+        B : 2d-array, square
+            A square adjacency matrix
+
+        seeds_A : 1d-array, shape (m , 1) where m <= number of nodes (default = [])
+            An array where each entry is an index of a node in `A`.
+
+        seeds_B : 1d-array, shape (m , 1) where m <= number of nodes (default = [])
+            An array where each entry is an index of a node in `B` The elements of
+            `seeds_A` and `seeds_B` are vertices which are known to be matched, that is,
+            `seeds_A[i]` is matched to vertex `seeds_B[i]`.
+
+        Returns
+        -------
+        perm_inds_ : 1-d array, some shuffling of [0, n_vert)
+            The optimal permutation indices to minimize the objective function
+        """
+        self.fit(A, B, seeds_A, seeds_B)
+        return self.nomination_list
+
+
+def ego(graph_adj_matrix, order, node, mindist=1):
+    # Note all nodes are zero based in this implementation, i.e the first node is 0
+    dists = [[node]]
+    for ii in range(1, order + 1):
+        clst = []
+        for nn in dists[-1]:
+            clst.extend(list(np.where(graph_adj_matrix[nn] == 1)[0]))
+        clst = list(set(clst))
+
+        # Remove all the ones that are closer (i.e thtat have already been included)
+        dists_conglom = []
+        for dd in dists:
+            dists_conglom.extend(dd)
+        dists_conglom = list(set(dists_conglom))
+
+        cn_proc = []
+        for cn in clst:
+            if cn not in dists_conglom:
+                cn_proc.append(cn)
+
+        dists.append(cn_proc)
+    ress = []
+
+    for ii in range(mindist, order + 1):
+        ress.extend(dists[ii])
+
+    return np.array(list(set(ress)))
```

Comment:
At least two of the above nested loops can be eliminated by pre-allocating arrays and then using numpy / scipy batch operations to fill them. This will drastically speed up function on large graphs.